karate-tests/src/test/java/patients/healthcareWorker/postPatient.feature

```diff
@@ -1,48 +1,138 @@
-            "regOrganisation.extension": "RWF",
-            "authorPersonID": "",
-            "authorSystemID": "230811201324",
-            "deathStatus": "",
-            "deceasedTime": "",
-            "overallUpdateMode": "create"
-        }
-    }
-    """
-    * path "Patient"
-    * request body
-    * method post
-    * status 201
-    * def nhsNumber = response.id
-    * def expectedResponse = read('classpath:stubs/patient/new-nhs-number-response-template.json')
-    * match response == expectedResponse
+  * def utils = call read('classpath:helpers/utils.feature')
+  * def faker = Java.type('helpers.FakerWrapper')
+  
+  # the same registeringAuthority object is included in every request
+  * def registeringAuthority = 
+  """
+  {
+    "regAuthorityType.code": "x",
+    "regAuthorityType.codeSystem": "2.16.840.1.113883.2.1.3.2.4.16.20",
+    "regOrganisation.root": "2.16.840.1.113883.2.1.4.3",
+    "regOrganisation.extension": "RWF",
+    "authorPersonID": "",
+    "authorSystemID": "230811201324",
+    "deathStatus": "",
+    "deceasedTime": "",
+    "overallUpdateMode": "create"
+  }
+  """
+
+  # the addressSchema used for the create responses is not the same as what's defined elsewhere, so 
+  # we'll use this schema until the schemas are harmonised...
+  * def addressSchema = 
+  """
+  {
+    "id": "#regex([0-9A-Z]+)",
+    "line": [
+        "#string"
+    ],
+    "period": {
+        "start": "#? utils.isTodaysDate(_)"
+    },
+    "postalCode": "#? utils.isValidPostalCode(_)",
+    "use": "#regex(home)"
+  }
+  """
+  
+  * def accessToken = karate.callSingle('classpath:patients/healthcareWorker/auth-redirect.feature').accessToken
+  * def requestHeaders = call read('classpath:patients/healthcareWorker/healthcare-worker-headers.js')
+  * configure headers = requestHeaders  
+  * url baseURL
+
+  # we have to chill out a bit between requests in case we trigger the spike alert policy :-(
+  * eval utils.sleep(4)
+
+  
+Scenario: Post patient - new patient
+  * def familyName = "Karate-test-" + utils.randomString(7)
+  * def givenName = "Zebedee"
+  * def prefix = "Mr"
+  * def gender = "male"
+  * def genderCode = "1"
+  * def birthDate = utils.randomBirthDate()
+  * def birthTime = birthDate.replaceAll("-","")
+  * def postalCode = faker.postalCode()
+  * def address = faker.streetAddress()
+  
+  * path "Patient"
+  * request read('classpath:patients/healthcareWorker/post-patient-request.json')
+  * method post
+  * status 201
+  * def nhsNumber = response.id
+  * def expectedResponse = read('classpath:stubs/patient/new-nhs-number-response-template.json')
+  * match response == expectedResponse
+  * match response.address[0].line[0] == address
+  * match response.address[0].postalCode == postalCode
+
+
+Scenario: Fail to create a record for a new patient, single demographics match found
+  # we rely on data that's already in the database for our existing record
```

Comment:
In this scenario, would it be better to send to post requests with the same details, expecting the first to return a `201/CREATED` response and the second a `200/SINGLE_MATCH`? Thus reducing our reliance on particular test data being in the environment?

Reply:
That was how I did it originally, and it's not hard to do, but we have another scenario to test the 201 response. Until we have control over the data, I think this is the best option - we minimise the amount of patients we create. If we find that the tests fail because data in veit07 is changing too much, then we can review, but for now I'm comfortable with this. We do a get request to make sure the data is as expected, so if we get a failure, I think it's clear what the issue is.

Reply:
👍